src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Created ->- Activating ->- Activated ->- Deactivating ->- Deactivated -------------->----------- Disposed
+ *                 |  |                                         |  |                                  |  |
+ *                 |  `------------------<----------------------'  `--->--- FailedToDeactivate --->---'  |
+ *                 `--------------------->--------- FailedToActivate ------------------>-----------------'
+ */
+
+enum LifeCycleState {
+  Created = 1,
+  Activating,
+  Activated,
+  Deactivating,
+  Deactivated,
+  Disposed,
+  FailedToActivate,
+  FailedToDeactivate,
+}
+
+const LifeCycleDescription = {
+  [LifeCycleState.Created]: 'has never activated before',
+  [LifeCycleState.Activating]: 'is activating',
+  [LifeCycleState.Activated]: 'is activated',
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
+  isActivated: boolean;
+  isDeactivating: boolean;
+  isDeactivated: boolean;
+  isErrored: boolean;
+  activate: () => Promise<void>;
+  deactivate: () => Promise<void>;
+  disposeActivateable: () => void;
+}
+
+export abstract class Activatable implements IActivatable {
+  private _state: LifeCycleState = LifeCycleState.Created;
+
+  public get isActivating() {
+    return this._state === LifeCycleState.Activating;
+  }
+
+  public get isActivated() {
+    return this._state === LifeCycleState.Activated;
+  }
+
+  public get isDeactivating() {
+    return this._state === LifeCycleState.Deactivating;
+  }
+
+  public get isDeactivated() {
+    return (
+      this._state === LifeCycleState.Created ||
+      this._state === LifeCycleState.Deactivated ||
+      this._state === LifeCycleState.Disposed
+    );
+  }
+
+  public get isErrored() {
+    return this._state === LifeCycleState.FailedToActivate || this._state === LifeCycleState.FailedToDeactivate;
+  }
+
+  public async activate() {
+    switch (this._state) {
+      case LifeCycleState.Activated:
+        return;
```

Comment:
Not sure, but I usually prefer to 'break' in such cases. Leaving it up to you.

Reply:
eh I usually prefer `return` to `break`. just a style thing.

Reply:
Yeah, just code style, can stay as is.